metaflow/plugins/kfp/kfp.py

```diff
@@ -28,35 +29,166 @@ def get_ordered_steps(graph):
-    TODO: The public docker is a copy of the internal docker image we were using (borrowed from aip-kfp-example). Check if any stage here may need to be further modified later.
     """
+    import subprocess
+    from collections import namedtuple
+
+    print("\n----------RUNNING: CODE DOWNLOAD from URL---------")
+    subprocess.call(["curl -o helloworld.py {}".format(code_url)], shell=True)
+
+    print("\n----------RUNNING: KFP Installation---------------")
+    subprocess.call(["pip3 install kfp"], shell=True) # TODO: Remove this once KFP is added to dependencies
+
+    print("\n----------RUNNING: METAFLOW INSTALLATION----------")
+    subprocess.call(["pip3 install --user --upgrade git+https://github.com/zillow/metaflow.git@state-integ-s3"],
+                    shell=True)
+
+    print("\n----------RUNNING: MAIN STEP COMMAND--------------")
+    S3_BUCKET = "s3://workspace-zillow-analytics-stage/aip/metaflow"
+    S3_AWS_ARN = "arn:aws:iam::170606514770:role/dev-zestimate-role"
+
+    define_s3_env_vars = 'export METAFLOW_DATASTORE_SYSROOT_S3="{}" && export METAFLOW_AWS_ARN="{}" '.format(S3_BUCKET,
+                                                                                                             S3_AWS_ARN)
+    define_username = 'export USERNAME="kfp-user"'
+    python_cmd = "python helloworld.py --datastore s3 --datastore-root {0} step {1} --run-id {2} " \
+                 "--task-id {3} --input-paths {2}/{4}/{5}".format(ds_root, step_name, run_id,
+                                                                  task_id, prev_step_name, prev_task_id)
+    final_run_cmd = f'{define_username} && {define_s3_env_vars} && {python_cmd}'
+
+    print("RUNNING COMMAND: ", final_run_cmd)
+    proc = subprocess.run(final_run_cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True)
+    proc_output = proc.stdout
+    proc_error = proc.stderr
+
+    step_output = namedtuple('StepOutput',
+                             ['ds_root', 'run_id', 'next_step', 'next_task_id', 'current_step', 'current_task_id'])
+
+    # END is the final step and no outputs need to be returned
+    if step_name.lower() == 'end':
+        print("_______________ FLOW RUN COMPLETE ________________")
+        return step_output('None', 'None', 'None', 'None', 'None', 'None')
+
+    # TODO: Check why MF echo statements are getting redirected to stderr
+    if len(proc_error) > 1:
+        print("_______________STDERR:_____________________________")
+        print(proc_error)
+
+    if len(proc_output) > 1:
+        print("_______________STDOUT:_____________________________")
+        print(proc_output)
+        # Read in the outputs (in the penultimate line) containing args needed for the next step.
+        # Note: Output format is: ['ds_root', 'run_id', 'next_step', 'next_task_id', 'current_step', 'current_task_id']
+        outputs = (proc_output.split("\n")[-2]).split()
```

Comment:
Guessing this output format is defined by Metaflow itself? Seems to me to be a somewhat flaky solution given I'm guessing the log output format is not meant to be a public interface we can rely on.

Obvious alternatives are to me:
1. Piping these details to a known/passed file location and reading (easy but ugly)
1. Instead of using subprocesses to call the step calling the code in the same python process and side-stepping the click CLI (much harder but more palatable in my opinion).

Reply:
With my latest changes, the need to read the outputs of one step to begin the next step is eliminated and so the `step_op_func` no longer does this :) (It's currently being used just once where we need to read the initial `run_id` that gets generated. There's a task to use the KFP run_id and so once that is implemented, this can be removed completely)